category/basic.lean

```diff
@@ -44,14 +58,34 @@ end
   (g <$> (x <*> y)) = ((∘) g) <$> x <*> y :=
 by simp [(pure_seq_eq_map _ _).symm]; simp [seq_assoc]
 
+private def mpartition_aux (x : α) : ulift bool → list α × list α → list α × list α
+ | ⟨ tt ⟩ (xs,ys) := (x::xs,ys)
+ | ⟨ ff ⟩ (xs,ys) := (xs,x::ys)
+
+def list.mpartition' (g : α → f (ulift bool)) : list α → f (list α × list α)
+ | [] := pure ([],[])
+ | (x :: xs) := mpartition_aux x <$> g x <*> list.mpartition' xs
+
+def list.mpartition {α : Type} {f : Type → Type v} [applicative f] [is_lawful_applicative f]
+  (g : α → f bool) :=
+list.mpartition' (λ x, ulift.up <$> g x)
+
+variables {m : Type u → Type v} [applicative m]
+def lift₂
+  {α₁ α₂ φ : Type u}
+  (f : α₁ → α₂ → φ)
+  (ma₁ : m α₁) (ma₂: m α₂) : m φ :=
+f <$> ma₁ <*> ma₂
+
 end applicative
 
 -- TODO: setup `functor_norm` for `monad` laws
 section monad
 variables {m : Type u → Type v} [monad m] [is_lawful_monad m]
 
 lemma map_bind (x : m α) {g : α → m β} {f : β → γ} : f <$> (x >>= g) = (x >>= λa, f <$> g a) :=
-by simp [bind_assoc, (∘), (bind_pure_comp_eq_map _ _ _).symm]
+by { rw [← bind_pure_comp_eq_map,bind_assoc],
+     simp [bind_pure_comp_eq_map], }
```

Comment:
Style: no final comma, also use `by tac1; tac2` instead of `by { tac1, tac2 }` if possible